include/circt/Dialect/LTL/LTLOps.td

```diff
@@ -0,0 +1,312 @@
+  let description = [{
+    If any of the `$inputs` is of type `!ltl.property`, the result of the op is
+    an `!ltl.property`. Otherwise it is an `!ltl.sequence`.
+  }];
+}
+
+def OrOp : AndLikeOp<"or"> {
+  let summary = "A disjunction of booleans, sequences, or properties.";
+  let description = [{
+    If any of the `$inputs` is of type `!ltl.property`, the result of the op is
+    an `!ltl.property`. Otherwise it is an `!ltl.sequence`.
+  }];
+}
+
+//===----------------------------------------------------------------------===//
+// Sequences
+//===----------------------------------------------------------------------===//
+
+def DelayOp : LTLOp<"delay", [Pure]> {
+  let arguments = (ins
+    LTLAnySequenceType:$input,
+    I64Attr:$delay,
+    OptionalAttr<I64Attr>:$length);
+  let results = (outs LTLSequenceType:$result);
+  let assemblyFormat = [{
+    $input `,` $delay (`,` $length^)? attr-dict `:` type($input)
+  }];
+  let hasFolder = 1;
+  let hasCanonicalizeMethod = 1;
+
+  let summary = "Delay a sequence by a number of cycles.";
+  let description = [{
+    Delays the `$input` sequence by the number of cycles specified by `$delay`.
+    The delay must be greater than or equal to zero. The optional `$length`
+    specifies during how many cycles after the initial delay the sequence can
+    match. Omitting `$length` indicates an unbounded but finite delay. For
+    example:
+
+    - `ltl.delay %seq, 2, 0` delays `%seq` by exactly 2 cycles. The resulting
+      sequence matches if `%seq` matches exactly 2 cycles in the future.
+    - `ltl.delay %seq, 2, 2` delays `%seq` by 2, 3, or 4 cycles. The resulting
+      sequence matches if `%seq` matches 2, 3, or 4 cycles in the future.
+    - `ltl.delay %seq, 2` delays `%seq` by 2 or more cycles. The number of
+      cycles is unbounded but finite, which means that `%seq` *has* to match at
+      some point, instead of effectively never occuring by being delayed an
+      infinite number of cycles.
+    - `ltl.delay %seq, 0, 0` is equivalent to just `%seq`.
+
+    #### Clocking
+
+    The cycle delay specified on the operation refers to a clocking event. This
+    event is not directly specified by the delay operation itself. Instead, the
+    [`ltl.clock`](#ltlclock-circtltlclockop) operation can be used to associate
+    all delays within a sequence with a clock.
+  }];
+}
+
+def ConcatOp : LTLOp<"concat", [Pure]> {
+  let arguments = (ins Variadic<LTLAnySequenceType>:$inputs);
+  let results = (outs LTLSequenceType:$result);
+  let assemblyFormat = [{
+    $inputs attr-dict `:` type($inputs)
+  }];
+  let hasFolder = 1;
+  let hasCanonicalizeMethod = 1;
+
+  let summary = "Concatenate sequences into a longer sequence.";
+  let description = [{
+    Concatenates all of the `$inputs` sequences one after another into one
+    longer sequence. The sequences are arranged such that the end time of the
+    previous sequences coincides with the start time of the next sequence. This
```

Comment:
The overlap is an important piece to make the concat operation truly have no delay. If the end of the LHS coincides with the beginning of the RHS, there is no need to describe what the delay between the LHS end and the RHS beginning is. If the concatenated sequences would be non-overlapping, the question would come up as to what that delay between the two is. Is it a clock cycle? Which clock? Or is it a time delay? The implication operator has the exact same problem, where in SVA the non-overlapping implication has terrible semantics and interact weirdly with zero-length sequences. The overlapping implication is trivial, since the two sequences overlap in a single point (`lhs <= x`, `rhs >= x`, so x is the single point of overlap; of Lebesgue measure zero!), and that point relates the timing of one squence with the other without the need to specify the amount of difference in the timing.